design/baremetal-operator/sub-states.md

```diff
@@ -0,0 +1,155 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost detailed activity
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes a new field to specify more precisely at which point in
+a BMH lifecycle it is.
+
+## Motivation
+
+The primary motivation is to enable operators an easier insight into what
+is going on with their nodes. This simplifies debugging and makes people feel
+better about long-running processes. It may also be helpful for UI tools
+to be able to display some sort of progress information.
+
+### Goals
+
+* Provide a way to know what action is being executed right now without
+  accessing Ironic API.
+
+### Non-Goals
+
+* Provide an access to the Ironic Node `provision_state` field.
+* Establish a direct one-to-one mapping between provision states and the
+  new field values.
+* Provide more granularity than what the Ironic API can report.
+* The values of the new field will not be a part of the API contract and can
+  change as a result of backend changes.
+* Define a non-empty value for each possible situation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator reasonably familiar with the provisioning process, I would like
+to know what action is being executed on a BMH right now.
+
+## Design Details
+
+* The `ProvisionStatus` structure is extended with a new string field
+  `CurrentActivity`.
+* The value of the field is re-calculated based on the result described below
+  on each reconciliation.
+
+### Implementation Details/Notes/Constraints
+
+The new field will be populated based on the following rules, where `Node` is
+an Ironic Node object. The rules are defined in pseudo-code.
+
+This proposal introduces a concept of "cleaning" with a narrower definition
+that in Ironic. While in Ironic "cleaning" refers to any pre-deployment
+actions, in our case "cleaning" will be used in the context of "cleaning
+disks".
+
+* IF `Node.provision_state` in "cleaning", "clean wait":
+
+  * IF `Node.clean_step` is null
+    AND `Node.target_provision_state` is "available":
+
+    Activity is `Preparing to clean disks`.
+
+  * IF `Node.clean_step` is null
+    AND `Node.target_provision_state` is "manageable":
+
+    Activity is `Preparing to configure host`.
+
+  * IF `Node.clean_step["interface"]` == "raid":
+
+    Activity is `Configuring RAID`.
+
+  * IF `Node.clean_step["interface"]` == "bios":
+
+    Activity is `Configuring firmware settings`.
+
+  * IF `Node.clean_step["step"]` in "erase\_devices",
+    "erase\_devices\_metadata":
+
+    Activity is `Cleaning disks`.
```

Comment:
```suggestion
    Activity is `Deleting disk metadata`.
```

Reply:
In theory, we can also enable full cleaning (this is what `erase_devices` is about). But okay, we can change this for now.